### DISCUSSION TITLE
the trait `Deserialize<'_>` is not implemented for `deadpool::managed::Pool<Manager>` for a custom perseus implementation.

### DISCUSSION THREAD

**Discussion by afidegnum** [framesurge/perseus#138] [General]
Hi, I have series of errors while trying to implement a few states from the `get_built_state` function trying to get the global state. 

I am trying to use `deadpool_postgres` alongside with if possible tokio or actix. but facing a series of challenges. among others are the error I posted above. 
Here is the entire error

```rs

error[E0308]: mismatched types
  --> src/global_state.rs:29:1
   |
29 | #[perseus::make_rx(AppStateRx)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `BasicPool`, found struct `BasicPoolRx`
   |
   = note: this error originates in the attribute macro `perseus::make_rx` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no method named `make_unrx` found for struct `BasicPool` in the current scope
  --> src/global_state.rs:29:1
   |
29 | #[perseus::make_rx(AppStateRx)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method not found in `BasicPool`
...
36 | pub struct BasicPool {
   | -------------------- method `make_unrx` not found for this
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `make_unrx`, perhaps you need to implement it:
           candidate #1: `MakeUnrx`
   = note: this error originates in the attribute macro `perseus::make_rx` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `deadpool::managed::Pool<Manager>: Serialize` is not satisfied
    --> src/global_state.rs:37:5
     |
37   |     pub basics: Pool,
     |     ^^^ the trait `Serialize` is not implemented for `deadpool::managed::Pool<Manager>`
     |
note: required by a bound in `configs::_::_serde::ser::SerializeStruct::serialize_field`
    --> /home/afidegnum/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.136/src/ser/mod.rs:1899:12
     |
1899 |         T: Serialize;
     |            ^^^^^^^^^ required by this bound in `configs::_::_serde::ser::SerializeStruct::serialize_field`

error[E0277]: the trait bound `deadpool::managed::Pool<Manager>: Deserialize<'_>` is not satisfied
    --> src/global_state.rs:37:5
     |
37   |     pub basics: Pool,
     |     ^^^ the trait `Deserialize<'_>` is not implemented for `deadpool::managed::Pool<Manager>`
     |
note: required by a bound in `next_element`
    --> /home/afidegnum/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.136/src/de/mod.rs:1725:12
     |
1725 |         T: Deserialize<'de>,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `next_element`

error[E0277]: the trait bound `deadpool::managed::Pool<Manager>: Deserialize<'_>` is not satisfied
    --> src/global_state.rs:37:5
     |
37   |     pub basics: Pool,
     |     ^^^ the trait `Deserialize<'_>` is not implemented for `deadpool::managed::Pool<Manager>`
     |
note: required by a bound in `next_value`
    --> /home/afidegnum/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.136/src/de/mod.rs:1864:12
     |
1864 |         V: Deserialize<'de>,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `next_value`
```
Can you please assist?

Here is[ the repo](https://github.com/afidegnum/myblog):

**Comment by arctic-hen7** [comment]
From the look of your code, you're trying to create a database pool at build time and put that into the global state. The problem is that any global or page state in Perseus must implement `Serialize` and `Deserialize`, it must be able to be converted to and from a string. This is so that state can be sent over the network to the browser.

Of course, you can't turn a database pool into a string, so you can't put it into the state. If you can, I'd advise getting what you need from the database at build time and then putting *that* into the state. If you need the database pool in the browser (which I don't think you do), you should have a separate server that accesses it on the browser's behalf, otherwise you'd be exposing the secret keys used to access the DB.

**Reply by afidegnum** [reply]
> "If you can, I'd advise getting what you need from the database at build time and then putting that into the state. If you need the database pool in the browser (which I don't think you do),"

How do i execute the database function at build time and pass its values into the states?

**Reply by arctic-hen7** [reply]
Well, that depends what you need from the database. I'm not sure of exactly how your system works, so I can't help you there. But anything you do need to send to the browser that can be sent as a string can just be generated in the build state function and returned as a part of the state.

**Comment by afidegnum** [comment]
Thanks, I also thought of that, and that's what I'm working on now. in that case, How can I ensure secure communication between perseus app and the auxiliary server? I'm considering using JWT what d you suggest?

**Reply by arctic-hen7** [reply]
JWTs should work for now, but there are a few problems with them that you may want to consider. Really any authentication strategy will work, it's just a matter of implementing it.

I'll have the docs on this part of the authentication written in the next few weeks, which should make things clearer in future.

**Reply by afidegnum** [reply]
Thanks
